### PR TITLE
DOC-1782: API documentation for searchable menu buttons with new example

### PR DESCRIPTION
**Pull request by ephox-mogran**
Related Ticket: DOC-1782

Description of Changes:
* Added documentation and an example for the new searchable toolbar menu buttons

Pre-checks:
- [x] Branch prefixed with `feature/6/` or `hotfix/6/`
- [x] ~`modules/ROOT/nav.adoc` has been updated (if applicable)~
- [x] Files has been included where required (if applicable)
- [x] ~Files removed have been deleted, not just excluded from the build (if applicable)~
- [x] (New product features only) Release Note added

Review:
- [x] Documentation Team Lead has reviewed

### CODE REVIEW

File: modules/ROOT/examples/live-demos/custom-toolbar-menu-button-search/index.js [hunk 1]
```diff
@@ -0,0 +1,53 @@
+tinymce.init({
```
Comment (ephox-mogran) [inline modules/ROOT/examples/live-demos/custom-toolbar-menu-button-search/index.js:1]: I wasn't intending to do an example, but I felt that my explanation of how it worked was pretty hard to follow, and an example would help. Plus with this livedemo system that has been setup, it wasn't as painful as I was expecting.

File: modules/ROOT/examples/live-demos/custom-toolbar-menu-button-search/index.js [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+        if (fetchContext.pattern.length > 0) {
+          callback([
+            {
+              type: 'menuitem',
+              text: `You searched for: "${fetchContext.pattern}"`,
+              onAction: () => editor.insertContent(`<strong>Inserted selected search result</strong>`)
+            }
+          ])
+        } else {
+          const items = [
```
Comment (ephox-mogran) [inline modules/ROOT/examples/live-demos/custom-toolbar-menu-button-search/index.js:27]: It might be confusing that I removed the nestedmenu items here, and it may cause readers to think that they aren't supported. I mainly simplified the example to focus on what was different about it. I'm not sure if we need to explicitly mention that nestedmenus are supported, or whether it is sufficient to just not mention that they aren't.

File: modules/ROOT/examples/live-demos/custom-toolbar-menu-button-search/index.js [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+    /* example, adding a toolbar menu button */
+    editor.ui.registry.addMenuButton('mybutton', {
+      text: 'My searchable button',
+      search: {
+        placeholder: 'What do you want to search for?'
+      },
+      fetch: (callback, fetchContext) => {
+        if (fetchContext.pattern.length > 0) {
+          callback([
+            {
```
Comment (ephox-mogran) [inline modules/ROOT/examples/live-demos/custom-toolbar-menu-button-search/index.js:20]: Perhaps an actual search would have been a better example. This example's intent though is to demonstrate that you can use pattern however you like.

File: modules/ROOT/pages/custom-menu-toolbar-button.adoc [hunk 4]
```diff
@@ -39,4 +40,25 @@ This example configures a toolbar menu button with the label `+My Button+` that
...
+
+The `+fetchContext+` is an object containing a single string property: `+pattern+`. If the toolbar menu button has not configured `search` to be active, then the `+pattern+` string will be empty.
+
+== Searchable menu button example and explanation
+
+include::partial$misc/admon-requires-6.2v.adoc[]
+
+The following is a simple toolbar menu button example, where searching has been configured:
+
+liveDemo::custom-toolbar-menu-button-search[tab="js"]
```
Comment (ephox-mogran) [inline modules/ROOT/pages/custom-menu-toolbar-button.adoc:58]: Initially I had the Searchable Menu header section above the first demo, but it just became really hard to follow, so I moved the whole section below the first demo so that all the searchable menu content is in one place. I'm not tied to it. @brian-forte change however you see fit.

File: modules/ROOT/pages/custom-menu-toolbar-button.adoc [hunk 5]
```diff
@@ -39,4 +40,25 @@ This example configures a toolbar menu button with the label `+My Button+` that
...
+
+This example configures a toolbar menu button with the label `+My searchable button+` that opens the specified menu when clicked. The menu will contain a search input field because `+search+` is not `+false+`. The input field's https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input#placeholder[placeholder attribute] will be `+What do you want to search for?+`.
```
Comment (ephox-mogran) [inline modules/ROOT/pages/custom-menu-toolbar-button.adoc:60]: Heads-up: I've probably been inconsistent when using `blah` vs `+blah+`.

File: modules/ROOT/pages/custom-menu-toolbar-button.adoc (hunk 4 above)
Comment (ephox-mogran) [reply modules/ROOT/pages/custom-menu-toolbar-button.adoc:58]: Note: also this demo requires 6.2 (as you would expect)

File: modules/ROOT/pages/custom-menu-toolbar-button.adoc [hunk 6]
```diff
@@ -12,10 +12,11 @@ For example: The table plugin's `+table+` toolbar button opens a menu similar to
 [cols="1,2,1,4",options="header"]
 |===
 |Name |Value |Requirement |Description
-|fetch |`+(success: (menu) => void) => void+` |required |Function that takes a callback which must be passed the list of options for the button's dropdown.
+|fetch |`+(success: (menu) => void, fetchContext) => void+` |required |Function that takes a callback and a `+fetchContext+` object. The callback function must be passed the list of options for the button's dropdown. The `+fetchContext+` object provides information which is useful for choosing which items should be passed to the callback. For more details on `fetchContext`, see xref:searchable-menu-buttons[Searchable Menu Buttons]
```
Comment (ephox-mogran) [inline modules/ROOT/pages/custom-menu-toolbar-button.adoc:15]: I looked in various places, and we didn't seem to have a standard way of specifying the types of objects passed into functions, so because that detail is necessary to use this function, I've just put a small amount of detail in the table, and then linked to a lower section that has more information. The lower section still doesn't explicitly state the types, though. As I said, other examples (https://www.staging.tiny.cloud/docs/tinymce/6/dialog-configuration/) aren't particularly consistent on how they detail the information in objects passed through to functions. In some cases, the information doesn't seem to be present at all.

The admonition about this being only in >= 6.2 was also moved to the linked section, instead of putting even more content in this table (as was discussed).

File: modules/ROOT/pages/custom-menu-toolbar-button.adoc [hunk 7]
```diff
@@ -39,4 +40,25 @@ This example configures a toolbar menu button with the label `+My Button+` that
 
 The `+fetch+` function is called when the toolbar menu button's menu is opened. It is a function that takes a callback and passes it an array of menu items to be rendered in the drop-down menu. This allows for asynchronous fetching of the menu items.
 
+[[searchable-menu-buttons]]
+== Searchable Menu Buttons
+
+include::partial$misc/admon-requires-6.2v.adoc[]
```
[the remark this answers is not in the document]
Comment (ephox-mogran) [reply modules/ROOT/pages/custom-menu-toolbar-button.adoc:46]: - [ ] ensure file is merged first (or add it to this PR)

File: modules/ROOT/examples/live-demos/custom-toolbar-menu-button-search/index.js (hunk 1 above)
Comment (MitchCTiny) [reply modules/ROOT/examples/live-demos/custom-toolbar-menu-button-search/index.js:1]: I'm not sure what our rules are around having multiple demos on the same page, but I believe the folder this demo is within should reflect the name of the page it is included on. Perhaps the `custom-menu-toolbar-button` page could be split into two pages, and as reflected in the nav by a heading with two sub pages, one for the simple menu button and one for the searchable menu button? This would be a question for @brian-forte .

File: modules/ROOT/pages/custom-menu-toolbar-button.adoc (hunk 7 above)
[the remark this answers is not in the document]
Comment (MitchCTiny) [reply modules/ROOT/pages/custom-menu-toolbar-button.adoc:46]: This file exists within the footnotes PR, you should be able to copy it to this PR. This was done in the 6.1 release and no conflicts occur (as far as I can remember).

File: modules/ROOT/examples/live-demos/custom-toolbar-menu-button-search/index.js (hunk 1 above)
Comment (lnewson) [reply modules/ROOT/examples/live-demos/custom-toolbar-menu-button-search/index.js:1]: We have multiple demos on the one page where it makes sense to do it like this. We do have the primary demo try and match the page name so it's easier to correlate where it makes sense to though. For example: https://github.com/tinymce/tinymce-docs/blob/e6829425aec9f3f5f54e6b4eb7d426a7ebc6de1f/modules/ROOT/pages/autocompleter.adoc#L195-L201

Comment (MitchCTiny) [reply modules/ROOT/examples/live-demos/custom-toolbar-menu-button-search/index.js:1]: Ah okay, thanks for clarifying.

File: modules/ROOT/examples/live-demos/custom-toolbar-menu-button-search/index.js (hunk 2 above)
Comment (MitchCTiny) [reply modules/ROOT/examples/live-demos/custom-toolbar-menu-button-search/index.js:27]: Might be confusing to include them in the same button, I think mentioning that they are supported should suffice. I feel that this fetch callback should include a few example search items though, rather than just a single action for anything searched.

File: modules/ROOT/pages/custom-menu-toolbar-button.adoc (hunk 6 above)
Comment (MitchCTiny) [reply modules/ROOT/pages/custom-menu-toolbar-button.adoc:15]: Since this option is used in all menu buttons, not just the searchable menu button, I don't think referencing the searchable menu buttons makes this clear. I wonder if stating that there are examples of this function in both live demos.

File: modules/ROOT/pages/custom-menu-toolbar-button.adoc [hunk 8]
```diff
@@ -39,4 +40,25 @@ This example configures a toolbar menu button with the label `+My Button+` that
...
+
+The button's menu can be configured to have an input field for searching, as well as its usual items. The presence of the input field is controlled by the `+search+` field in the xref:options[options]. The `+search+` field can be a `+boolean+` or an `+object+` containing a single optional string `placeholder`. By default, `+search+` is false. If `+search+` is not false, the button's menu will contain an input field with any specified placeholder text. As the user types into this field, the `+fetch+` function will be called with the text in the input field passed back as part of `+fetchContext+`. The `+fetch+` function is responsible for using this `+fetchContext+` to determine which items to pass to its `+success+` callback.
+
+The `+fetchContext+` is an object containing a single string property: `+pattern+`. If the toolbar menu button has not configured `search` to be active, then the `+pattern+` string will be empty.
+
+== Searchable menu button example and explanation
```
Comment (MitchCTiny) [inline modules/ROOT/pages/custom-menu-toolbar-button.adoc:52]:
```suggestion
=== Searchable menu button example and explanation
```
This currently has the same heading level as `Searchable Menu Buttons`

**Review by MitchCTiny — APPROVED** [review]

File: modules/ROOT/examples/live-demos/custom-toolbar-menu-button-search/index.js [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+              type: 'menuitem',
+              text: `You searched for: "${fetchContext.pattern}"`,
+              onAction: () => editor.insertContent(`<strong>Inserted selected search result</strong>`)
+            }
+          ])
```
Comment (tiny-ben-tran) [inline modules/ROOT/examples/live-demos/custom-toolbar-menu-button-search/index.js:25]:
```suggestion
          ]);
```

Missing a semi colon :)

File: modules/ROOT/pages/custom-menu-toolbar-button.adoc (hunk 5 above)
Comment (tiny-ben-tran) [inline modules/ROOT/pages/custom-menu-toolbar-button.adoc:60]:
```suggestion
This example configures a toolbar menu button with the label `+My searchable button+` that opens the specified menu when clicked. The menu will contain a search input field because `+search+` is not `+false+`. The input field's https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input#placeholder[placeholder attribute] will be `+Type...+`.
```

**Review by tiny-ben-tran — APPROVED** [review]

File: modules/ROOT/examples/live-demos/custom-toolbar-menu-button-search/index.html [hunk 10]
```diff
@@ -0,0 +1,20 @@
...
+    <li>Our <a href="{{site-url}}/tinymce/6/">documentation</a> is a great resource for learning how to configure TinyMCE.</li>
+    <li>Have a specific question? Try the <a href="{{communitysupporturl}}" target="_blank" rel="noopener"><code>{{prodnamecode}}</code> tag at Stack Overflow</a>.</li>
+    <li>We also offer enterprise grade support as part of <a href="{{pricingpage}}">TinyMCE premium plans</a>.</li>
+  </ul>
+
+  <h2>Found a bug?</h2>
+  <p>If you think you have found a bug please create an issue on the <a href="https://github.com/tinymce/tinymce/issues">GitHub repo</a> to report it to the developers.</p>
+
+  <h2>Finally ...</h2>
+  <p>Don't forget to check out our other product <a href="http://www.plupload.com" target="_blank">Plupload</a>, your ultimate upload solution featuring HTML5 upload support.</p>
```
[the remark this answers is not in the document]
Comment (MitchCTiny) [reply modules/ROOT/examples/live-demos/custom-toolbar-menu-button-search/index.html:17]: Plupload is a single product, and likely just copy-pasted from a different demo

File: modules/ROOT/examples/live-demos/custom-toolbar-menu-button-search/index.js (hunk 2 above)
Comment (ephox-mogran) [reply modules/ROOT/examples/live-demos/custom-toolbar-menu-button-search/index.js:27]: Yeah, it's the difference between showing a typical use case versus showing how the button works. At the moment, it's just showing how the button works. Searching is the typical use case, but it is possible to use it for other things.

Comment (ephox-mogran) [reply modules/ROOT/examples/live-demos/custom-toolbar-menu-button-search/index.js:27]: See: https://github.com/tinymce/tinymce-docs/pull/2520#discussion_r976460331

File: modules/ROOT/pages/custom-menu-toolbar-button.adoc (hunk 7 above)
[the remark this answers is not in the document]
Comment (ephox-mogran) [reply modules/ROOT/pages/custom-menu-toolbar-button.adoc:46]: Yeah, that was the intent. That someone else would merge it first. I have it locally, but didn't see the need to add it to the PR.

File: modules/ROOT/pages/custom-menu-toolbar-button.adoc (hunk 6 above)
Comment (ephox-mogran) [reply modules/ROOT/pages/custom-menu-toolbar-button.adoc:15]: I wonder if the `Searchable Menu Buttons` part should just be called something closer to `FetchContext`. Let's clear this up with the team. This page differs from some of the others, which has made it hard to apply precedent and consistency.

File: modules/ROOT/pages/custom-menu-toolbar-button.adoc (hunk 8 above)
Comment (ephox-mogran) [reply modules/ROOT/pages/custom-menu-toolbar-button.adoc:52]: Yeah, that was intentional, because the other demo has this level. I think we probably just need to have a team discussion about how we want this file setup.

File: modules/ROOT/pages/custom-menu-toolbar-button.adoc (hunk 6 above)
Comment (ephox-mogran) [reply modules/ROOT/pages/custom-menu-toolbar-button.adoc:15]: We'll address this later as part of a "Cleanup Searchable Menu docs page"

File: modules/ROOT/pages/custom-menu-toolbar-button.adoc [hunk 11]
```diff
@@ -39,4 +40,25 @@ This example configures a toolbar menu button with the label `+My Button+` that
 
 The `+fetch+` function is called when the toolbar menu button's menu is opened. It is a function that takes a callback and passes it an array of menu items to be rendered in the drop-down menu. This allows for asynchronous fetching of the menu items.
 
+[[searchable-menu-buttons]]
+== Searchable Menu Buttons
```
Comment (ephox-mogran) [inline modules/ROOT/pages/custom-menu-toolbar-button.adoc:44]: - [ ] inconsistent capitalisation

Comment (ephox-mogran) [reply modules/ROOT/pages/custom-menu-toolbar-button.adoc:44]:
```suggestion
== Searchable menu buttons
```

**Review by brian-forte — APPROVED** [review]
